src/backends/postgresql/session.cpp

```diff
@@ -67,6 +67,16 @@ void postgresql_session_backend::connect(
                          : "SET extra_float_digits = 2",
         "Cannot set extra_float_digits parameter");
 
+    PGresult* res = PQexec(conn, "SHOW search_path");
+    if (PQresultStatus(res) == PGRES_TUPLES_OK)
+    {
+        if (PQntuples(res) > 0)
+        {
+            schema_name_ = PQgetvalue(res, 0, 0);
```

Comment:
How does this work? Search path is something like `"$user", public` by default for me, which can't be used as the schema name...

Reply:
You are right, I lost track of how this works. Search path in PostgreSQL can hold 0 or more comma separated schema names, if it holds 0, public will be the default, if it holds more than one, I am actually not sure how it works. I assume (this must indeed be verified) that for DDL creates it will use the first schema to create into, and for any other operation it'll keep searching until the first table it finds.

No matter how it works it of course changes how PR should alter the current behavior that I still believe is not great for anybody using a schema which is not public.

Reply:
Yes, it uses the first existing schema for creating the new tables and any existing schema for lookup, see [the docs](https://www.postgresql.org/docs/current/ddl-schemas.html#DDL-SCHEMAS-PATH).

Reply:
So, the PR is currently broken - obviously - and I'm going to fix it.

However, how I should fix it is the big question.

Using the search path is what is used everywhere else in SOCI not because it's explicitly implemented like that, but it's implicitly like that due to the use of the underlying RDBMS.
So for the sake of consistency I would use that, and that will change how the metadata API works in SOCI. (Hopefully to the better)

The question is, how should it actually work? I want the current state to remain, ie. if you don't have a search_path I want to use the public schema - and maybe $user too - this is not how it has been working historically, but maybe the original implementation was not meant for anything but public? - I've been trying to figure that one out.
If $user should be used as well, then I need to implement handling of multiple schemas which is not currently implemented. Currently, looking up tables just returns a table of table names. But maybe the schema name should be there as well? However, that will break backward compatibility...

Going the opposite direction, aiming at keeping backward compatibility only one schema can be used with the metadata. I.e. public without the search path and only the first schema on the search path.

Please let me hear your thought.

Reply:
I don't think the existing query can work for Postgres, i.e. it really should be `where table_schema in list-of-schemas-in-the-search-path` instead of just simple equality. Perhaps Postgres has some nicer way of writing this condition, but the idea would be to use the same logic which is used for unqualified table names lookup.

Which brings me to another point: we should support qualified table names too and in this case there shouldn't be any condition involving `table_schema` at all, i.e. I suspect this is broken too currently.

In any case, this definitely needs new tests (even if they're Postgres-specific).

Reply:
Maybe I'm trying to force a kube into a round hole.

As a user of SOCI I really would prefer if the <code>details::prepare_temp_type</code> returned by <code>session::prepare_table_names()</code> has the same layout for any RDBMS (i.e. not returning both schema and table name for one DB and then only table name for everything else). 

My preferred option would be to have both fields in the output - I think it is cleaner. However, in the hunt for backward compatibility I would suggest the following:

- <code>session::prepare_table_names()</code> returns one field, for PostgreSQL this is a qualified table name. I.e. "schema.table".
- <code>postgresql_session_backend::get_table_names_query()</code> must be implemented to select the "TABLE_NAME" column such that it is created from both the schema_name and table_name and the schema_name must be used in the where clause to select all schemas in the search path, as discussed above.
- <code>details::prepare_temp_type session::prepare_column_descriptions(std::string & table_name)</code> would then always look for a dot and if it exists parse the schema- and table name to give it to the prepared statement as <code>:s</code> and <code>:t</code> respectively. The default statement in <code>session_backend::get_column_description_query()</code> could then make use of this <code>:s</code> in the query, avoiding SQL injection. For <code>table_name</code> input without a schema the default value would simply be "public". Other implementations of <code>get_column_description_query()</code> doesn't have to use the <code>:s</code>. Maintaining backward compatibility.

This solution would allow [schema.]table syntax input to <code>details::prepare_temp_type session::prepare_column_descriptions(std::string & table_name)</code>. However, the tables returned by <code>session::prepare_table_names()</code> will still depend on what is in the search path.

Reply:
Sorry, maybe we need to rewind and actually define what problem exactly are we trying to solve here. I thought it was just about getting metadata of the tables in non-default schema but it actually seems something different? If so, could you please define what exactly it is, i.e. what do you need to do that doesn't work right now?

Reply:
The problem is that the current implementation hardcodes the public schema. This is hardcoded in both get_table_names_query() and get_column_description_query().

This doesn't make sense in PostgreSQL and even more so in MySQL. These are the DBs I'm familiar with - but it could be a problem for other DBs too.

For PostgreSQL I propose to use the search path to get the schemas that should be used for the search of table names.

However, as the same table names may exist in multiple schemas so want to return these in their qualified form and I want to also update the prepare_column_descriptions to be able to get retrieve the metadata for these tables.

Reply:
> Are you really sure about it? As long as unqualified names are always interpreted in the same way, I don't see any need to qualify them. Although it could be more clear, admittedly.

When having multiple schemas you can end up in this situation (obviously it will only happen ever so often):

```
postgres=# select table_schema || '.' || table_name as "TABLE_NAME" from information_schema.tables where table_schema in ('a', 'b',  'c');
 TABLE_NAME
-------------
 a.soci_test
 b.soci_test
 c.soci_test
(3 rows)
```

Which is the reason why, for the PostgreSQL implementation, I would return the qualified names as in the query suggested above.

Reply:
In a select statement in PostgreSQL it will work like this, sure.

This write-up mostly serves to convince myself that it can be done. 
To mimic the PostgreSQL behavior in the metadata API, which is indeed sensible, I believe <code>details::prepare_temp_type session::prepare_column_descriptions(std::string & table_name) const</code> and the underlying PostgreSQL can be changed as follows:

Currently <code>details::prepare_temp_type session::prepare_column_descriptions(std::string & table_name) const</code> looks like this:

```
details::prepare_temp_type session::prepare_column_descriptions(std::string & table_name)
{
    ensureConnected(backEnd_);

    return prepare << backEnd_->get_column_descriptions_query(), use(table_name, "t");
}
```

I already had the idea that I would change it to also have a sanitizing function returning a struct of schema and table name like:

```
details::prepare_temp_type session::prepare_column_descriptions(std::string & table_name)
{
    ensureConnected(backEnd_);

    struct table_data = getTableData(table_name);

    return prepare << backEnd_->get_column_descriptions_query(), use(table_data.schema, "s"), use(table_data.name, "t");
}
```

When no schema is present the value of <code>table_data.schema</code> would be <code>NULL</code>. The query for PostgreSQL would then be (Please notice I'm using an example in which I assume the search path to hold $user - being 'soci_user' - and 'public'):

```
select 
    column_name as \"COLUMN_NAME\",
    data_type as \"DATA_TYPE\",
    character_maximum_length as \"CHARACTER_MAXIMUM_LENGTH\",
    numeric_precision as \"NUMERIC_PRECISION\",
    numeric_scale as \"NUMERIC_SCALE\",
    is_nullable as \"IS_NULLABLE\"
from
    information_schema.columns
where
    table_name = :t
    and
    table_schema in ('soci-user', 'public')
    and 
    case
        when :s is not NULL THEN table_schema = :s
        else true
    end
order by
    case table_schema
        when 'soci-user' then 1
        when 'public' then 2
        else 3
    end
limit 1;
```

The interesting blocks which hold values that I can't figure out how to store as :-notation are the in-clause which holds the tables of the search path (I keep them in order but it's irrelevant) which must have the same length as the length of the search path and the order by-clause which also holds the values of the search path and once again must have that length to maintain the order of the result - i.e. the soci-user schema must come before the public schema when no schema is present in :s.

As I wrote in the beginning I'm writing this to convince myself it can be done, however, please let me know if you find anything that seems to be overly complicated.